File: src/main/java/com/articulo/model/Articulo.java
package com.articulo.model;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.List;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.venta.model.TipoPago;

import jakarta.persistence.CascadeType;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.FetchType;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.OneToMany;
import jakarta.persistence.Table;
import jakarta.persistence.UniqueConstraint;
import jakarta.transaction.Transactional;
import jakarta.validation.constraints.DecimalMin;
import jakarta.validation.constraints.Digits;
import jakarta.validation.constraints.NotNull;

@Entity
@Table(uniqueConstraints = {@UniqueConstraint(columnNames = {"codigo", "negocio"})})
public class Articulo extends BaseItem{
	
	//uso Integer para poder diferencias null de 0, a diferencia del tipo primitivo int
	@Column
	private Integer stock;
	
	@Column(nullable = false)
	@DecimalMin(value = "0", inclusive = true, message = "El precio minimo es de 0.00.")
	@Digits(integer=8, fraction=2, message = "El precio maximo es de 99,999,999.99.")
	protected BigDecimal precioCompra;
	
	//relacion con Combo
	@JsonIgnore
	@OneToMany(fetch = FetchType.LAZY)
	@JoinColumn(name="id_articulo")
	private List<ArticuloCombo> combo;
	
	//relacion con Grupo
	@JsonIgnoreProperties({"hibernateLazyInitializer", "handler"})
	@ManyToOne(optional = false, fetch = FetchType.EAGER)
	@JoinColumn(name="id_grupo_articulo")
	@NotNull(message = "El articulo debe estar asociado a un grupo.")
	private GrupoArticulo grupoArticulo;
	
    @OneToMany(mappedBy = "articulo", fetch = FetchType.EAGER, cascade = CascadeType.ALL, orphanRemoval = true)
    private List<Imagen> imagenes = new ArrayList<>();
	
	public Articulo() {}

	public Articulo editar(Articulo a) {
		//this.codigo = a.getCodigo();
		this.descripcion = a.getDescripcion();
		this.grupoArticulo = a.getGrupoArticulo();
		this.nombre = a.getNombre();
		this.precioCompra = a.getPrecioCompra();
		this.stock = a.getStock();
		removeAllImagenes();
		addAllImagenes(a.getImagenes());
		return this;
	}

	
    public void addImagen(Imagen imagen) {
        imagenes.add(imagen);
        imagen.setArticulo(this);
    }

    public void removeImagen(Imagen imagen) {
        imagenes.remove(imagen);
        imagen.setArticulo(null);
    }
    
    @Transactional
    public void removeAllImagenes() {
        for (Imagen imagen : new ArrayList<>(imagenes)) {
            removeImagen(imagen);
        }
        imagenes.clear();
        /*
    	List<Imagen> lista= this.imagenes;
    	for(Imagen i : lista) {
    		removeImagen(i);
    	}
    	*/
    }
    
    @Transactional
    public void addAllImagenes(List<Imagen> imagenes) {
    	for(Imagen i : imagenes) {
    		addImagen(i);
    	}
    }

	public List<Imagen> getImagenes() {
		return imagenes;
	}

	public void setImagenes(List<Imagen> imagenes) {
		this.imagenes = imagenes;
	}

	public Integer getStock() {
		return stock;
	}

	public void setStock(Integer stock) {
		this.stock = stock;
	}
	
	public List<ArticuloCombo> getCombo() {
		return combo;
	}

	public void setCombo(List<ArticuloCombo> combo) {
		this.combo = combo;
	}

	public GrupoArticulo getGrupoArticulo() {
		return grupoArticulo;
	}

	public void setGrupoArticulo(GrupoArticulo grupoArticulo) {
		this.grupoArticulo = grupoArticulo;
	}
	
	@Override
	public BigDecimal getPrecioPorTipoPago(TipoPago tipoPago) {
		
		BigDecimal porcentajeRecargo = new BigDecimal(grupoArticulo.getPorcentajeRecargoPorTipoPago(tipoPago));
		BigDecimal multiplicador = porcentajeRecargo.divide(new BigDecimal(100)).add(new BigDecimal(1));
		return this.precioCompra.multiply(multiplicador);
	}

	@Override
	public void reducirStock(Integer cantidad) {
		this.stock = this.stock - cantidad;
	}
	
	@Override
	public String getType() {
		this.type = "Item";
		return type;
	}

	public BigDecimal getPrecioCompra() {
		return precioCompra;
	}

	public void setPrecioCompra(BigDecimal precioCompra) {
		this.precioCompra = precioCompra;
	}
	
}

File: angular/src/app/articulo/model/articulo.ts
import { DetalleVenta } from "../../venta/model/detalle-venta";
import { TipoPago } from "../../venta/model/tipo-pago";
import { ArticuloCombo } from "./articulo-combo";
import { BaseItem } from "./base-item";
import { GrupoArticulo } from "./grupo-articulo";
import { Imagen } from "./imagen";
import { Negocio } from "./negocio";
import { Type } from "./type";

export class Articulo implements BaseItem{


    id?:number;
	nombre?:string;
	descripcion?:string;
	type?:string;
	codigo?:string;
	negocio?:Negocio;
	detallesVentas?:DetalleVenta[];
	precioCompra? : number;

    imagenes:Imagen[]=[];

    stock? : number;
    grupoArticulo? : GrupoArticulo;
    combo? : ArticuloCombo[];

    constructor(){}

    setArticulo(obj:any){
        this.id=obj.id;
        this.nombre=obj.nombre;
        this.descripcion=obj.descripcion;
        this.type=obj.type;
        this.codigo=obj.codigo;
        this.negocio=obj.negocio;
        this.detallesVentas=obj.detallesVentas;
        this.precioCompra=obj.precioCompra;
        this.stock=obj.stock;
        this.grupoArticulo=obj.grupoArticulo;
        this.combo=obj.combo;
    }

    calcularPrecioPorTipoPago(tipoPago: TipoPago): number {
        console.log(tipoPago);
        switch (tipoPago) {
            case 0:
                if (this.grupoArticulo?.porcentajeRecargoSinIVA && this.precioCompra) {
                    return Number((((this.grupoArticulo?.porcentajeRecargoSinIVA /100) + 1) * this.precioCompra).toFixed(2));
                }
                break;
            case 1:
                if (this.grupoArticulo?.porcentajeRecargoConIVA && this.precioCompra) {
                    return Number((((this.grupoArticulo?.porcentajeRecargoConIVA /100) + 1) * this.precioCompra).toFixed(2));
                }
                break;
            case 2:
                if (this.grupoArticulo?.porcentajeRecargoTarjeta && this.precioCompra) {
                    return Number((((this.grupoArticulo?.porcentajeRecargoTarjeta /100) + 1) * this.precioCompra).toFixed(2));
                }
                break;
            case 3:
                if (this.grupoArticulo?.porcentajeRecargoTarjeta3Cuotas && this.precioCompra) {
                    return Number((((this.grupoArticulo?.porcentajeRecargoTarjeta3Cuotas /100) + 1) * this.precioCompra).toFixed(2));
                }
                break;
            case 4:
                if (this.grupoArticulo?.porcentajeRecargoTarjeta6Cuotas && this.precioCompra) {
                    return Number((((this.grupoArticulo?.porcentajeRecargoTarjeta6Cuotas /100) + 1) * this.precioCompra).toFixed(2));
                }
                break;
        }
        return 0;
    }
    
}
